pkg/xds/experimental/client/client.go

```diff
@@ -0,0 +1,621 @@
+}
+
+func (c *Client) upsertAndDeleteMissing(upsertedResources map[string]map[string]proto.Message) error {
+	var errs error
+	for _, typeUrl := range []string{envoy.ListenerTypeURL, envoy.ClusterTypeURL, envoy.EndpointTypeURL} {
+		curr, ok := upsertedResources[typeUrl]
+		if !ok {
+			c.log.Debug("no changes", "typeUrl", typeUrl)
+			continue
+		}
+		var deletedResources []string
+		var err error
+		if typeUrl == envoy.EndpointTypeURL {
+			deletedResources, err = c.deletedEndpoints(upsertedResources)
+		} else {
+			deletedResources, err = c.deletedResources(typeUrl, curr)
+		}
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+		c.log.Debug("cache TX", "typeUrl", typeUrl, "upserted", slices.Collect(maps.Keys(curr)), "deleted", deletedResources)
+		ver, updated, _ := c.cache.TX(typeUrl, curr, deletedResources)
+		c.log.Debug("cache TX", "typeUrl", typeUrl, "ver", ver, "updated", updated)
+	}
+	return errs
+}
+
+func (c *Client) deletedEndpoints(upsertedResources map[string]map[string]proto.Message) ([]string, error) {
+	// Handle partial update:
+	// Endpoint is deleted only if Clusters were part of the update.
+	clusterNames, ok := upsertedResources[envoy.ClusterTypeURL]
+	if !ok {
+		return nil, nil
+	}
+	old, err := c.getAllResources(envoy.EndpointTypeURL)
+	if err != nil {
+		// In version 1.14 GetResources doesn't return any error for these arguments.
+		return nil, fmt.Errorf("get old resources: %w", err)
+	}
+	deletedResources := make([]string, 0, len(old.ResourceNames))
+	// Handle partial update:
+	// The corresponding Cluster must be missing from a non-empty update.
+	for _, name := range old.ResourceNames {
+		if _, ok := clusterNames[name]; ok {
+			continue
+		}
+		deletedResources = append(deletedResources, name)
+	}
+	return deletedResources, nil
+}
+
+func (c *Client) deletedResources(typeUrl string, curr map[string]proto.Message) ([]string, error) {
+	old, err := c.getAllResources(typeUrl)
+	if err != nil {
+		// In version 1.14 GetResources doesn't return any error for these arguments.
+		return nil, fmt.Errorf("get old resources: %w", err)
+	}
+	deletedResources := make([]string, 0, len(old.ResourceNames))
+	for _, name := range old.ResourceNames {
+		if _, ok := curr[name]; ok {
+			continue
+		}
+		deletedResources = append(deletedResources, name)
+	}
+	return deletedResources, nil
+}
+
+func (c *Client) sendACK(ctx context.Context, stream discoverypb.AggregatedDiscoveryService_StreamAggregatedResourcesClient, resp *discoverypb.DiscoveryResponse, upsertedResources map[string]map[string]proto.Message) error {
+	var errs error
+	for typeUrl, typeResources := range upsertedResources {
+		resourceNames := slices.Collect(maps.Keys(typeResources))
+		req := &discoverypb.DiscoveryRequest{
+			Node:          c.node,
+			VersionInfo:   resp.GetVersionInfo(),
+			ResponseNonce: resp.GetNonce(),
+			TypeUrl:       typeUrl,
+			ResourceNames: resourceNames,
+		}
+		select {
+		case <-ctx.Done():
```

Comment:
Checking of ctx.Done() seems fishy to me.
Does stream.Send() block? Does it respond to cancelation?
If the answer is no, then the checking of ctx.Done() is race condition.

Reply:
yes, `stream.Send` blocks. yes, it responds to cancelation